crates/providers/src/heart.rs

```diff
@@ -1,23 +1,203 @@
-//! Block Hearbeat and Transaction Watcher
+//! Block hearbeat and pending transaction watcher.
 
+use crate::Provider;
+use alloy_network::Network;
 use alloy_primitives::{B256, U256};
 use alloy_rpc_types::Block;
-use alloy_transport::{utils::Spawnable, TransportErrorKind, TransportResult};
+use alloy_transport::{utils::Spawnable, Transport, TransportErrorKind, TransportResult};
 use futures::{stream::StreamExt, FutureExt, Stream};
 use std::{
     collections::{BTreeMap, HashMap},
     fmt,
     future::Future,
+    marker::PhantomData,
     time::{Duration, Instant},
 };
 use tokio::{
     select,
     sync::{mpsc, oneshot, watch},
 };
 
-/// A configuration object for watching for transaction confirmation.
+/// A builder for configuring a pending transaction watcher.
+///
+/// # Examples
+///
+/// Send and wait for a transaction to be confirmed 2 times, with a timeout of 60 seconds:
+///
+/// ```no_run
+/// # async fn example<N: alloy_network::Network>(provider: impl alloy_providers::Provider<N>, tx: N::TransactionRequest) -> Result<(), Box<dyn std::error::Error>> {
+/// // Send a transaction, and configure the pending transaction.
+/// let builder = provider.send_transaction(tx)
+///     .await?
+///     .with_confirmations(2)
+///     .with_timeout(Some(std::time::Duration::from_secs(60)));
+/// # let builder = builder.with_provider(&provider); // TODO
+/// // Register the pending transaction with the provider.
+/// let pending_transaction = builder.register().await?;
```

Comment:
`watch` does `register().await?.await`, we need to expose both because creating a pending transaction is async too due to use of tokio channels, vs this step not existing in ethers